// @ionic/app-scripts/dist/bundle.js
"use strict";
var util_1 = require('./util');
var path_1 = require('path');
var fs_extra_1 = require('fs-extra');
var os_1 = require('os');
function bundle(context, options, rollupConfig, useCache) {
    if (useCache === void 0) { useCache = true; }
    context = util_1.generateContext(context);
    options = util_1.generateBuildOptions(options);
    var logger = new util_1.Logger("bundle " + (options.isProd ? 'prod' : 'dev'));
    // bundle the app then create create css
    return runBundle(context, options, rollupConfig, useCache).then(function () {
        return logger.finish();
    }).catch(function (reason) {
        return logger.fail(reason);
    });
}
exports.bundle = bundle;
function bundleUpdate(event, path, context, options, useCache) {
    if (useCache === void 0) { useCache = true; }
    var logger = new util_1.Logger("bundle " + (options.isProd ? 'prod' : 'dev') + " update");
    return runBundle(context, options, null, useCache).then(function () {
        return logger.finish();
    }).catch(function (err) {
        return logger.fail(err);
    });
}
exports.bundleUpdate = bundleUpdate;
function runBundle(context, options, rollupConfig, useCache) {
    var taskInfo = (options.isProd) ? ROLLUP_PROD_TASK_INFO : ROLLUP_TASK_INFO;
    rollupConfig = util_1.fillConfigDefaults(context, rollupConfig, taskInfo);
    rollupConfig.dest = path_1.join(context.buildDir, rollupConfig.dest);
    if (useCache) {
        // tell rollup to use a previous bundle as its starting point
        rollupConfig.cache = bundleCache;
    }
    if (!rollupConfig.onwarn) {
        rollupConfig.onwarn = createOnWarnFn();
    }
    // bundle the app then create create css
    var rollup = require('rollup').rollup;
    return rollup(rollupConfig).then(function (bundle) {
        // set the module files used in this bundle
        // this reference can be used elsewhere in the build (sass)
        context.moduleFiles = bundle.modules.map(function (m) { return m.id; });
        // async cache all the module paths so we don't need
        // to always bundle to know which modules are used
        setModulePathsCache(context.moduleFiles);
        // Cache our bundle for later use
        bundleCache = bundle;
        // write the bundle
        return bundle.write(rollupConfig);
    });
}
function getModulePathsCache() {
    // sync get the cached array of module paths (if they exist)
    var modulePaths = null;
    var modulesCachePath = getModulesPathsCachePath();
    try {
        modulePaths = fs_extra_1.readJsonSync(modulesCachePath, { throws: false });
        util_1.Logger.debug("Cached module paths: " + (modulePaths && modulePaths.length) + ", " + modulesCachePath);
    }
    catch (e) {
        util_1.Logger.debug("Cached module paths not found: " + modulesCachePath);
    }
    return modulePaths;
}
exports.getModulePathsCache = getModulePathsCache;
function setModulePathsCache(modulePaths) {
    // async save the module paths for later lookup
    var modulesCachePath = getModulesPathsCachePath();
    fs_extra_1.outputJson(modulesCachePath, modulePaths, function (err) {
        if (err) {
            util_1.Logger.error("Error writing module paths cache: " + err);
        }
        util_1.Logger.debug("Cached module paths: " + (modulePaths && modulePaths.length) + ", " + modulesCachePath);
    });
}
function getModulesPathsCachePath() {
    // make a unique tmp directory for this project's module paths cache file
    var cwd = process.cwd().replace(/-|:|\/|\\|\.|~|;|\s/g, '').toLowerCase();
    if (cwd.length > 40) {
        cwd = cwd.substr(cwd.length - 40);
    }
    return path_1.join(os_1.tmpdir(), cwd, 'modulepaths.json');
}
// used to track the cache for subsequent bundles
var bundleCache = null;
function clearCachedModule(id) {
    if (bundleCache) {
        var cachedModule = bundleCache.modules.find(function (m) { return m.id === id; });
        if (cachedModule) {
            var index = bundleCache.modules.indexOf(cachedModule);
            if (index > -1) {
                bundleCache.modules.splice(index, 1);
                return true;
            }
        }
    }
    return false;
}
exports.clearCachedModule = clearCachedModule;
function createOnWarnFn() {
    var previousWarns = {};
    return function onWarningMessage(msg) {
        if (msg in previousWarns) {
            return;
        }
        previousWarns[msg] = true;
        if (!(IGNORE_WARNS.some(function (warnIgnore) { return msg.indexOf(warnIgnore) > -1; }))) {
            util_1.Logger.warn("rollup: " + msg);
        }
    };
}
var IGNORE_WARNS = [
    'keyword is equivalent to',
    'plugin (\'ng-template\') was used to transform files'
];
var ROLLUP_TASK_INFO = {
    fullArgConfig: '--rollup',
    shortArgConfig: '-r',
    envConfig: 'ionic_rollup',
    defaultConfigFilename: 'rollup.config'
};
var ROLLUP_PROD_TASK_INFO = {
    fullArgConfig: '--rollup',
    shortArgConfig: '-r',
    envConfig: 'ionic_rollup',
    defaultConfigFilename: 'rollup.prod.config'
};
