// @ionic/app-scripts/dist/template.js
"use strict";
var util_1 = require('./util');
var bundle_1 = require('./bundle');
var path_1 = require('path');
var fs_1 = require('fs');
var sass_1 = require('./sass');
function templateUpdate(event, path, context, options) {
    path = path_1.join(context.rootDir, path);
    var logger = new util_1.Logger('templateUpdate');
    return runTemplateUpdate(event, path, context, options).then(function () {
        // congrats, we did it!
        return logger.finish();
    }).catch(function (err) {
        return logger.fail(err, 'templateUpdate failed' + (err.message ? ': ' + err.message : ''));
    });
}
exports.templateUpdate = templateUpdate;
function runTemplateUpdate(event, path, context, options) {
    if (event === 'change') {
        // just a change event, see if this html file has a component in the same directory
        // doing this to prevent an unnecessary TS compile and bundling without cache if it was just a HTML change
        var componentFile = getSourceComponentFile(path, context);
        if (bundle_1.clearCachedModule(componentFile)) {
            // we successfully found the compiled JS file and cleared it from the bundle cache
            return bundle_1.bundleUpdate(event, path, context, options, true);
        }
    }
    // not sure how it changed, just do a full rebuild without the bundle cache
    return bundle_1.bundleUpdate(event, path, context, options, false).then(function () {
        return sass_1.sassUpdate(event, path, context, options, true);
    });
}
function getSourceComponentFile(htmlFilePath, context) {
    var rtn = null;
    try {
        var changedHtmlFile = path_1.basename(htmlFilePath);
        var componentDir = path_1.dirname(htmlFilePath);
        var filePaths = fs_1.readdirSync(componentDir);
        var match = void 0;
        for (var i = 0; i < filePaths.length; i++) {
            var filePath = filePaths[i];
            if (util_1.isTsFilename(filePath)) {
                // found a .ts file in this same directory
                // open it up and see if it's a component
                // and see if it has a template url with the same filename
                var tsComponentFile = path_1.join(componentDir, filePath);
                var source = fs_1.readFileSync(path_1.join(componentDir, filePath)).toString();
                if ((match = COMPONENT_REGEX.exec(source)) !== null) {
                    if ((match = TEMPLATE_URL_REGEX.exec(match[0])) !== null) {
                        var componentHtmlFile = path_1.basename(match[1].replace(/\'|\"|\`/g, '').trim());
                        if (changedHtmlFile === componentHtmlFile) {
                            rtn = getCompiledJsFile(tsComponentFile, context);
                            break;
                        }
                    }
                }
            }
        }
    }
    catch (e) {
        util_1.Logger.error(e);
    }
    return rtn;
}
function getCompiledJsFile(tsComponentFile, context) {
    var jsCompiledFile = tsComponentFile.replace(context.srcDir, context.tmpDir);
    return jsCompiledFile.substr(0, jsCompiledFile.length - 2) + 'js';
}
function inlineTemplate(options, source, sourcePath) {
    var componentDir = path_1.parse(sourcePath).dir;
    var match;
    var rewrite;
    var didRewrite = false;
    var sourceScan = source;
    options.directoryMaps = options.directoryMaps || {
        '.tmp': 'src'
    };
    while ((match = COMPONENT_REGEX.exec(sourceScan)) !== null) {
        rewrite = match[0].replace(TEMPLATE_URL_REGEX, function (m, urlValue) {
            if (urlValue.indexOf('\'') > -1 || urlValue.indexOf('"') > -1 || urlValue.indexOf('`') > -1) {
                didRewrite = true;
                return replaceTemplate(options, componentDir, urlValue);
            }
            return urlValue;
        });
        if (didRewrite) {
            source = source.replace(match[0], rewrite);
        }
        sourceScan = sourceScan.substring(match.index + match[0].length);
    }
    if (didRewrite) {
        return source;
    }
    return null;
}
exports.inlineTemplate = inlineTemplate;
function replaceTemplate(options, componentDir, urlValue) {
    return urlValue.replace(HTML_PATH_URL_REGEX, function (match, quote, filePath) {
        return inlineSourceWithTemplate(options, componentDir, filePath);
    });
}
function inlineSourceWithTemplate(options, componentDir, filePath) {
    var rtn = "templateUrl: '" + filePath + "'";
    try {
        var htmlPath = path_1.join(componentDir, filePath);
        for (var k in options.directoryMaps) {
            htmlPath = htmlPath.replace(path_1.sep + k + path_1.sep, path_1.sep + options.directoryMaps[k] + path_1.sep);
        }
        var htmlContent = fs_1.readFileSync(htmlPath).toString();
        htmlContent = htmlContent.replace(/\r|\n/g, '\\n');
        htmlContent = htmlContent.replace(/\'/g, '\\\'');
        rtn = "template: '" + htmlContent + "'";
    }
    catch (e) {
        console.error("Error reading template file, \"" + filePath + "\": " + e);
    }
    return rtn;
}
var COMPONENT_REGEX = /Component\s*?\(\s*?({([\s\S]*?)}\s*?)\)/m;
var TEMPLATE_URL_REGEX = /templateUrl\s*:(.*)/;
var HTML_PATH_URL_REGEX = /(['"])((?:[^\\]\\\1|.)*?)\1/g;
