// @ionic/app-scripts/dist/transpile.js
"use strict";
var util_1 = require('./util');
var path_1 = require('path');
function transpile(context) {
    context = util_1.generateContext(context);
    var logger = new util_1.Logger('transpile');
    return transpileApp(context).then(function () {
        return logger.finish();
    }).catch(function (reason) {
        return logger.fail(reason);
    });
}
exports.transpile = transpile;
function transpileUpdate(event, path, context) {
    return transpile(context);
}
exports.transpileUpdate = transpileUpdate;
function transpileApp(context, transpileConfig) {
    transpileConfig = util_1.fillConfigDefaults(context, transpileConfig, TRANSPILE_TASK_INFO);
    var srcFile = path_1.join(context.buildDir, transpileConfig.source);
    var destFile = path_1.join(context.buildDir, transpileConfig.destFileName);
    return transpile6To5(context, srcFile, destFile);
}
exports.transpileApp = transpileApp;
function transpile6To5(context, srcFile, destFile) {
    return new Promise(function (resolve, reject) {
        var spawn = require('cross-spawn');
        var tscCmd = path_1.join(context.rootDir, 'node_modules', '.bin', 'tsc');
        var tscCmdArgs = [
            '--out', destFile,
            '--target', 'es5',
            '--allowJs',
            '--sourceMap',
            srcFile
        ];
        var hadAnError = false;
        var ls = spawn(tscCmd, tscCmdArgs);
        ls.stdout.on('data', function (data) {
            util_1.Logger.info(data);
        });
        ls.stderr.on('data', function (data) {
            util_1.Logger.error("transpile error: " + data);
            hadAnError = true;
        });
        ls.on('close', function (code) {
            if (hadAnError) {
                reject("Transpiling from ES6 to ES5 encountered an error");
            }
            else {
                resolve();
            }
        });
    });
}
exports.transpile6To5 = transpile6To5;
var TRANSPILE_TASK_INFO = {
    fullArgConfig: '--tsc',
    shortArgConfig: '-tsc',
    envConfig: 'ionic_transpile',
    defaultConfigFilename: 'transpile.config'
};
